relay/nakama/siwe/siwe.go

```diff
@@ -0,0 +1,276 @@
+	now := time.Now()
+	obj.Nonces = slices.DeleteFunc(obj.Nonces, func(details *nonceWindow) bool {
+		return now.After(details.ExpiredAt)
+	})
+
+	if err = setNoncesStorageObject(ctx, nk, signerAddress, obj); err != nil {
+		return false, eris.New("failed to write new nonces back to DB")
+	}
+	return true, nil
+}
+
+// GenerateNewSIWEMessage generates an SIWE Message that can be signed and used to authenticate a user.
+func GenerateNewSIWEMessage(signerAddress string) (*GenerateResult, error) {
+	options := makeOptions()
+	nonce, err := makeNonce()
+	if err != nil {
+		return nil, err
+	}
+	msg, err := siwe.InitMessage(DefaultDomain, signerAddress, DefaultURI, nonce, options)
+	if err != nil {
+		return nil, eris.Wrap(err, "failed to init siwe message")
+	}
+
+	return &GenerateResult{
+		SIWEMessage: msg.String(),
+	}, nil
+}
+
+// setNonceStorageObject saves the given nonce storage object to Nakama's storage layer. If the version in the storage
+// object is out of date, an error will be returned.
+func setNoncesStorageObject(
+	ctx context.Context,
+	nk runtime.NakamaModule,
+	signerAddress string,
+	obj *nonceStorageObj,
+) error {
+	bz, err := json.Marshal(obj)
+	if err != nil {
+		return eris.Wrap(err, "failed to marshal nonces")
+	}
+
+	writes := []*runtime.StorageWrite{
+		{
+			Collection: siweMessageCollection,
+			Key:        signerAddress,
+			UserID:     utils.AdminAccountID,
+			Value:      string(bz),
+			Version:    obj.Version,
+		},
+	}
+	if _, err = nk.StorageWrite(ctx, writes); err != nil {
+		return eris.Wrap(err, "failed to save nonces to storage")
+	}
+	return nil
+}
+
+// getNonceStorageObject loads the nonces associated with the given signer address.
+func getNonceStorageObject(
+	ctx context.Context,
+	nk runtime.NakamaModule,
+	signerAddress string,
+) (
+	*nonceStorageObj, error,
+) {
+	read := []*runtime.StorageRead{
+		{
+			Collection: siweMessageCollection,
+			Key:        signerAddress,
+			UserID:     utils.AdminAccountID,
+		},
+	}
+	objs, err := nk.StorageRead(ctx, read)
+	if err != nil {
+		return nil, eris.Wrap(err, "failed to read valid nonces from DB")
+	} else if len(objs) == 0 {
+		// No existing storage object was found, so return a storage object with no nonces and no version
+		return &nonceStorageObj{
+			// When this is later saved back to the DB, it will only be successful it the stoage object
+			// doesn't already exist in the DB.
+			Version: "*",
+		}, nil
+	}
+
+	var nonceObj nonceStorageObj
+	if err := json.Unmarshal([]byte(objs[0].GetValue()), &nonceObj); err != nil {
+		return nil, eris.Wrap(err, "failed to unmarshal nonces")
+	}
+	nonceObj.Version = objs[0].GetVersion()
+	return &nonceObj, nil
+}
+
+func makeOptions() map[string]any {
+	now := time.Now().UTC()
+	return map[string]any{
+		"issuedAt":       now.Format(time.RFC3339),
+		"expirationTime": now.Add(DefaultTimeout).Format(time.RFC3339),
+		"statement":      DefaultStatement,
+	}
+}
+
+func makeNonce() (string, error) {
```

Comment:
nit: this should be `generateNonce` so as to not be confused with a regular constructor

Reply:
done